### PR TITLE
feat(streams): toTransformStream

### PR DESCRIPTION
**Pull request by ayame113**
part of #2208

Introduce a `generatorToStream` that transforms the generator function into a TransformStream, as described in https://github.com/denoland/deno_std/issues/2208#issuecomment-1122175133.

```ts
import { readableStreamFromIterable, generatorToStream } from "https://deno.land/std@$STD_VERSION/streams/mod.ts";

const readable = readableStreamFromIterable([0, 1, 2])
  .pipeThrough(generatorToStream(async function* (src) {
    for await (const chunk of src) {
      yield chunk * 100;
    }
  }));

for await (const chunk of readable) {
  console.log(chunk);
}
// output: 0, 100, 200
```

### CODE REVIEW

**Comment by crowlKats** [comment]
Havent looked into this closely yet, however, how does this differ from https://github.com/denoland/deno_std/blob/main/streams/conversion.ts#L174 (with the exception that this returns a transformstream instead of a readablestream)?

**Comment by ayame113** [comment]
It's basically similar, but this helps users create custom TransformStream classes.

```ts
class MyCustumTransformStream implements TransformStream {
  readonly writable: WritableStream<number>;
  readonly readable: ReadableStream<number>;
  constructor() {
    const transformer = generatorToStream(this.#transformFunc);
    this.writable = transformer.writable;
    this.readable = transformer.readable;
  }
  async *#transformFunc(src: ReadableStream<number>) {
    for await (const chunk of src) {
      yield chunk * 100;
    }
  }
}
```

**Comment by nayeemrmn** [comment]
I don't think the name `generatorToStream()` makes sense, generator is another name for the iterable iterator returned by the generator function.

**Comment by crowlKats** [comment]
also on top of that, `ToStream` is too generic. it should be `ToTransformStream`

**Comment by ayame113** [comment]
It is certainly so.
`transformStreamFromGeneratorFunction` is a long but accurate name, so is it better?

File: streams/conversion.ts [hunk 1]
```diff
@@ -196,6 +196,78 @@ export function readableStreamFromIterable<T>(
   });
 }
 
+/**
+ * Convert the generator function into a TransformStream.
+ *
+ * ```ts
+ * import { readableStreamFromIterable, transformStreamFromGeneratorFunction } from "https://deno.land/std@$STD_VERSION/streams/mod.ts";
```
Comment (bartlomieju) [inline streams/conversion.ts:203]: This should probably use relative import like other APIs above

**Review by bartlomieju — COMMENTED** [review]
The name of this API is a mouthful. Maybe we could use `transformStreamFromGenerator` instead?

File: streams/conversion.ts [hunk 2]
```diff
@@ -196,6 +196,78 @@ export function readableStreamFromIterable<T>(
...
+        if (typeof iterator.throw == "function") {
+          try {
+            await iterator.throw(reason);
+          } catch {
+            /* `iterator.throw()` always throws on site. We catch it. */
+          }
+        }
+        await readable.cancel(reason);
+      },
+    }, readableStrategy),
```
Comment (kt3k) [inline streams/conversion.ts:267]: This part looks similar to `readableStreamFromIterable` except some error handlings. How about updating `readableStreamFromIterable` to this implementation and using it here?

Comment (ayame113) [reply streams/conversion.ts:267]: I think it is possible to take the common parts of the two functions and integrate them into a another internal function.
Is this what you meant?

```ts
export function readableStreamFromIterable<T>(
  iterable: Iterable<T> | AsyncIterable<T>,
): ReadableStream<T> {
  return internalReadableStreamFromIterable(iterable);
}

export function transformStreamFromGenerator<I, O>(
  transformer: (src: ReadableStream<I>) => Iterable<O> | AsyncIterable<O>,
  writableStrategy?: QueuingStrategy<I>,
  readableStrategy?: QueuingStrategy<O>,
): TransformStream<I, O> {
  const {
    writable,
    readable,
  } = new TransformStream<I, I>(undefined, writableStrategy);

  const iterable = transformer(readable);
  return {
    writable,
    readable: internalReadableStreamFromIterable(iterable, readableStrategy, readable),
  };
}

function internalReadableStreamFromIterable<T>(
  iterable: Iterable<T> | AsyncIterable<T>,
  readableStrategy?: QueuingStrategy<T>,
  upstream?: ReadableStream,
): ReadableStream<T> {
  // implementation
}
```

File: streams/conversion.ts (hunk 1 above)
Comment (ayame113) [reply streams/conversion.ts]: Thank you for your review. Has been updated.

**Comment by nayeemrmn** [comment]
> The name of this API is a mouthful. Maybe we could use `transformStreamFromGenerator` instead?

@bartlomieju 

> I don't think the name `generatorToStream()` makes sense, generator is another name for the iterable iterator returned by the generator function.

Honestly I think we should consider naming this `transformStream()` and renaming `readableStreamFromIterable()` to `readableStream()`. The thing they convert from is the basic JS representation of each (iterable for readable stream, (iterable -> iterable) for transform stream).

**Comment by ayame113** [comment]
Regarding a function name, I would like to follow the decisions of the core team.

File: streams/conversion.ts (hunk 2 above)
Comment (kt3k) [reply streams/conversion.ts:267]: I imagined like that, but now I see it's a little confusing (I missed `readable` dependency in `internalReadableStreamFromIterable` 😓  )

The current implementation looks just fine to me.

**Comment by kt3k** [comment]
I'm in favor of @nayeemrmn's proposal (`transformStream()`). This utility seems useful for many cases. We can encourage the use of it by giving the short name.

What do you think? @crowlKats @bartlomieju

**Comment by crowlKats** [comment]
overall I am in favour, though I worry that having a function with the same name as a class (with change of casing) can get quite confusing for people

**Comment by kt3k** [comment]
> I worry that having a function with the same name as a class (with change of casing) can get quite confusing for people

Then how about `toTransformStream`?

**Comment by crowlKats** [comment]
yea, that sounds fine to me

**Comment by ayame113** [comment]
`toTransformStream` sounds good.👍 Has been updated.

**Review by kt3k — APPROVED** [review]
LGTM

We can still iterate on the name. Any further suggestions/opinions/discussions are welcome

**Review by bartlomieju — APPROVED** [review]

**Pull request merged**